packages/hardhat-core/test/builtin-tasks/compile.ts

```diff
@@ -844,4 +844,96 @@ Read about compiler configuration at https://hardhat.org/config
+
+  describe("project with files importing dependencies", function () {
+    useFixtureProject("compilation-contract-with-deps");
+    useEnvironment();
+
+    it("should not remove the build-info if it is still referenced by an external library", async function () {
+      await this.env.run("compile");
+
+      const pathContractA = path.join("contracts", "A.sol");
+      let contractA = fsExtra.readFileSync(pathContractA, "utf-8");
+      contractA = contractA.replace("contract A", "contract B");
+      fsExtra.writeFileSync(pathContractA, contractA, "utf-8");
+
+      /**
+       * The _validArtifacts variable is not cleared when running the compile
+       * task twice in the same process, leading to an invalid output. This
+       * issue is not encountered when running the task from the CLI as each
+       * command operates as a separate process. To resolve this, the private
+       * variable should be cleared after each run of the compile task.
+       */
+      // eslint-disable-next-line @typescript-eslint/dot-notation
+      (this.env.artifacts as any)["_validArtifacts"] = [];
+
+      await this.env.run("compile");
+
+      // asserts
+      const buildInfoPathB = path.join(
+        "artifacts",
+        "contracts",
+        "A.sol",
+        "B.dbg.json"
+      );
+      assertBuildInfoExists(buildInfoPathB);
+      const buildInfoPathConsole = path.join(
+        "artifacts",
+        "dependency",
+        "contracts",
+        "console.sol",
+        "console.dbg.json"
+      );
+      assertBuildInfoExists(buildInfoPathConsole);
+    });
+
+    it("should not remove the build-info if it is still referenced by another local contract", async function () {
+      await this.env.run("compile");
+
+      const pathContractC = path.join("contracts", "C.sol");
+      let contractC = fsExtra.readFileSync(pathContractC, "utf-8");
+      contractC = contractC.replace("contract C", "contract D");
+      fsExtra.writeFileSync(pathContractC, contractC, "utf-8");
+
+      /**
+       * The _validArtifacts variable is not cleared when running the compile
+       * task twice in the same process, leading to an invalid output. This
+       * issue is not encountered when running the task from the CLI as each
+       * command operates as a separate process. To resolve this, the private
+       * variable should be cleared after each run of the compile task.
+       */
+      // eslint-disable-next-line @typescript-eslint/dot-notation
+      (this.env.artifacts as any)["_validArtifacts"] = [];
+
+      await this.env.run("compile");
+
+      // asserts
+      const buildInfoPathC = path.join(
+        "artifacts",
+        "contracts",
+        "C.sol",
+        "D.dbg.json"
+      );
+      assertBuildInfoExists(buildInfoPathC);
+      const buildInfoPathE = path.join(
+        "artifacts",
+        "contracts",
+        "E.sol",
+        "E.dbg.json"
+      );
+      assertBuildInfoExists(buildInfoPathE);
+    });
+
+    afterEach(() => {
```

Comment:
I think I'd slightly prefer to do this in the tests themselves, because they are the ones doing this change so it makes sense that they also do the cleanup. An `afterEach` that performs a cleanup of something that wasn't done in a `beforeEach` feels a bit weird to me.

I understand that this means the files won't be reverted if the test fails, but I'm ok with that trade-off.

Reply:
Yeah, my main worry was something failing during the test and breaking future runs. But it's ok if you're good with the trade-off.